frontend/react/answers/samcrider.md

```diff
@@ -0,0 +1,23 @@
+## QUESTIONS
+
+### What is a prop?
+
+Props are the information that you pass to a JSX tag. For example: className and src. Components can pass props down to their children.
+
+### What is a hook? Can you explain what the built in hooks are, and when you would use them?
+
+Hooks are JS functions that employ stateful logic. React comes with several types of built-in hooks. These include state, context, ref, and effect hooks. State hooks store user input and that input can be recalled easily here. I would use a stateful hook to track an index because I could directly manipulate the index using useState. Context hooks allow for information to be passed down from a distant parent without using props. Ref hooks hold values that aren't rendered. These are important for working with non-react systems like the DOM. Finally, effect hooks are used to connect a react componen with an external system. An example would be a useEffect hook that performs an action upon an event being triggered.
+
+### What is a component? Why are components important?
+
+A component is a Javascript function that returns JSX. Components are important because they are the building blocks of applications and allow for cleaner, more readable code.
+
+### What is JSX?
+
+JSX is what is returned in components. It is basically HTML but you can write JS in curly braces in it.
```

Comment:
It is syntactically similar to* HTML, but quite different than HTML. Simple JSX is compiled by the React virtual DOM almost directly into HTML.

Reply:
It seems that jsx is compiled into Javascript at build time.

Reply:
Yeah, that's sort of correct (compiled is the wrong word for JSX -> HTML). Behaviorally JSX is pretty different than HTML (more than just writing JS in curly braces), although it is syntactically similar (XML-like).

This is actually a pretty interesting rabbithole to go down, but React uses a ["Virtual Dom"](https://legacy.reactjs.org/docs/faq-internals.html) (VDOM) in order to mirror the browsser-native HTML DOM. React Components are stored as nodes on the VDOM, and rendering happens using react-fiber, which recursively traverses the VDOM and "unravels" or converts React components from JSX to HTML almost 1:1.

So JSX is translated into standard JS (JSX is non-standard, which is why it will throw an error in a .js or .ts file vs .jsx or .tsx), and that JS is used to construct HTML blocks which are appended to the DOM by react fiber.

There's a really good explanation [here](https://dev.to/ktmouk/how-does-react-18-work-inside-1c8l)!

Reply:
oof